63_CUDAInterop/main.cpp

```diff
@@ -22,81 +20,115 @@ The start of the main function starts like in most other example. We ask the
+		system::IApplicationFramework(_localInputCWD, _localOutputCWD, _sharedInputCWD, _sharedOutputCWD) {}
 
 	smart_refctd_ptr<CCUDAHandler> cudaHandler;
 	smart_refctd_ptr<CCUDADevice> cudaDevice;
 
 	IQueue* queue;
 
-	std::array<smart_refctd_ptr<ICPUBuffer>, 2> cpubuffers;
-	std::array<smart_refctd_ptr<CCUDASharedMemory>, 3> mem = {};
-	smart_refctd_ptr<CCUDASharedSemaphore> cusema;
+	// CPU memory which we fill with random numbers between [-1,1] that will be copied to corresponding cudaMemory
+	std::array<smart_refctd_ptr<ICPUBuffer>, 2> cpuBufs;
+	// CUDA resources that we input to the kernel 'vectorAdd_kernel.cu'
+	// Kernel writes to cudaMemories[2] which we later use to export and read on nabla side
+	std::array<smart_refctd_ptr<CCUDASharedMemory>, 3> cudaMemories = {};
+	// A semaphore created in CUDA which will be imported into Nabla to help sync between the CUDA kernel and and Nabla device to host transfer
+	smart_refctd_ptr<CCUDASharedSemaphore> cudaSemaphore;
 
-	smart_refctd_ptr<IGPUBuffer> importedbuf, stagingbuf, stagingbuf2;
-	smart_refctd_ptr<IGPUImage> importedimg;
-	smart_refctd_ptr<ISemaphore> sema;
-	smart_refctd_ptr<IGPUCommandPool> commandPool;
-	smart_refctd_ptr<IGPUCommandBuffer> cmd;
+	// our Buffer that is bound to cudaMemories[2]
+	smart_refctd_ptr<IGPUBuffer> importedBuf;
+	// our Image that is also bound to cudaMemories[2]
+	smart_refctd_ptr<IGPUImage> importedImg;
+
+	// host visible buffers that we use to copy from the resources above after CUDA kernel is done writing
+	smart_refctd_ptr<IGPUBuffer> stagingBufs[2];
+
+	// Nabla semaphore that aliases the cudaSemaphore above
+	smart_refctd_ptr<ISemaphore> importedSemaphore;
 
+	smart_refctd_ptr<IGPUCommandPool> commandPool;
+	smart_refctd_ptr<IGPUCommandBuffer> cmd[2];
 
 	bool onAppInitialized(smart_refctd_ptr<ISystem>&& system) override
 	{
-		// Remember to call the base class initialization!
-		if (!base_t::onAppInitialized(std::move(system)))
-			return false;
-
+        // Remember to call the base class initialization!
+        if (!device_base_t::onAppInitialized(smart_refctd_ptr(system)))
+            return false;
+        if (!asset_base_t::onAppInitialized(std::move(system)))
+            return false;
+        
 		{
-			auto& limits = m_physicalDevice->getLimits();
-			if (!limits.externalMemoryWin32 || !limits.externalFenceWin32 || !limits.externalSemaphoreWin32)
-				return logFail("Physical device does not support the required extensions");
-			
-			cudaHandler = CCUDAHandler::create(system.get(), smart_refctd_ptr<ILogger>(m_logger));
-			assert(cudaHandler);
+			cudaHandler = CCUDAHandler::create(m_system.get(), smart_refctd_ptr<ILogger>(m_logger));
+			if (!cudaHandler) return logFail("Could not create a CUDA handler!");
 			cudaDevice = cudaHandler->createDevice(smart_refctd_ptr_dynamic_cast<CVulkanConnection>(m_api), m_physicalDevice);
+			if (!cudaDevice) return logFail("Could not create a CUDA Device!");
```

Comment:
1. initialize the Asset Base first
2. create your CUDA handler
3. override `MonoDeviceApplication::filterDevices` to only return devices that can be found within the `cudaHandler`'s list of device by matching UUIDs
4. now `device_base_t` will either return a correct Logical Device over a CUDA capable Physical Device or nothing at all

this kinda matters if you had an AMD/Intel GPU in one PCIE slot, and Nvidia in another, without a `filterDevices` override you might end up with the Logical Device created on the AMD one and example faling enven though it could proceed

Reply:
this is what https://github.com/Devsh-Graphics-Programming/Nabla-Examples-and-Tests/pull/90#discussion_r1442273763 was about